File: src/use-cases/bot/DeleteBot.js
/**
 * DeleteBot Use Case
 * Deletes a bot user
 */

const logger = require('../../../logger');

class DeleteBot {
    /**
     * @param {IUserRepository} userRepository - User repository
     */
    constructor(userRepository) {
        this.userRepository = userRepository;
    }

    /**
     * Execute the use case
     * @param {Object} request - Delete bot request
     * @param {string} request.botId - Bot user ID to delete
     * @returns {Promise<Object>} Deletion result
     */
    async execute({ botId }) {
        try {
            // Validate input
            if (!botId || typeof botId !== 'string') {
                throw new Error('Bot ID is required');
            }

            // Find bot
            const bot = await this.userRepository.findById(botId);
            if (!bot) {
                throw new Error('Bot not found');
            }

            // Verify it's actually a bot
            if (!bot.isBot()) {
                throw new Error('User is not a bot - cannot delete human users via this endpoint');
            }

            // Delete bot
            const deleted = await this.userRepository.delete(botId);

            if (!deleted) {
                throw new Error('Failed to delete bot');
            }

            logger.info('Bot deleted', {
                botId,
                username: bot.username
            });

            return {
                success: true,
                deletedBotId: botId
            };
        } catch (error) {
            logger.error('Delete bot error', {
                botId,
                error: error.message
            });

            throw error;
        }
    }
}

module.exports = DeleteBot;

File: zapzap-rust/src/application/bot/delete_bot.rs
//! Delete a bot user. Port of `src/use-cases/bot/DeleteBot.js`.

use std::sync::Arc;

use crate::application::bot::BotAdminError;
use crate::domain::repositories::UserRepository;

/// Delete bot use case
pub struct DeleteBot {
    user_repo: Arc<dyn UserRepository>,
}

impl DeleteBot {
    pub fn new(user_repo: Arc<dyn UserRepository>) -> Self {
        Self { user_repo }
    }

    /// Deletes the bot and returns its id; a human user is refused
    pub async fn execute(&self, bot_id: &str) -> Result<String, BotAdminError> {
        if bot_id.is_empty() {
            return Err(BotAdminError::Invalid("Bot ID is required".into()));
        }
        let bot = self
            .user_repo
            .find_by_id(bot_id)
            .await?
            .ok_or_else(|| BotAdminError::Invalid("Bot not found".into()))?;
        if !bot.is_bot() {
            return Err(BotAdminError::Invalid(
                "User is not a bot - cannot delete human users via this endpoint".into(),
            ));
        }
        self.user_repo.delete(bot_id).await?;
        tracing::info!("Bot deleted: {} ({})", bot.username, bot_id);
        Ok(bot_id.to_string())
    }
}
